Tools/cases_generator/optimizer_generator.py

```diff
@@ -147,19 +158,89 @@ def goto_label(self, goto: Token, label: Token, storage: Storage) -> None:
         self.out.emit(goto)
         self.out.emit(label)
 
+
+class OptimizerConstantEmitter(OptimizerEmitter):
+    def __init__(self, out: CWriter, labels: dict[str, Label], uop: Uop):
+        super().__init__(out, labels)
+        # Replace all outputs to point to their stackref versions.
+        overrides = {
+            outp.name: self.emit_stackref_override for outp in uop.stack.outputs
+        }
+        self._replacers = {**self._replacers, **overrides}
+
+    def emit_stackref_override(
+        self,
+        tkn: Token,
+        tkn_iter: TokenIterator,
+        uop: CodeSection,
+        storage: Storage,
+        inst: Instruction | None,
+    ) -> bool:
+        self.out.emit(tkn)
+        self.out.emit("_stackref ")
+        return True
+
+def write_uop_pure_evaluation_region_header(
+    uop: Uop,
+    out: CWriter,
+    stack: Stack,
+) -> None:
+    emitter = OptimizerConstantEmitter(out, {}, uop)
+    emitter.emit("if (\n")
+    assert len(uop.stack.inputs) > 0, "Pure operations must have at least 1 input"
+    for inp in uop.stack.inputs[:-1]:
+        emitter.emit(f"sym_is_const(ctx, {inp.name}) &&\n")
+    emitter.emit(f"sym_is_const(ctx, {uop.stack.inputs[-1].name})\n")
+    emitter.emit(') {\n')
+    # Declare variables, before they are shadowed.
+    for inp in uop.stack.inputs:
+        if inp.used:
+            emitter.emit(f"{type_name(inp)}{inp.name}_sym = {inp.name};\n")
+    # Shadow the symbolic variables with stackrefs.
+    for inp in uop.stack.inputs:
+        if inp.used:
+            emitter.emit(f"{stackref_type_name(inp)}{inp.name} = sym_get_const_as_stackref(ctx, {inp.name}_sym);\n")
+    # Rename all output variables to stackref variant.
+    for outp in uop.stack.outputs:
+        assert not outp.is_array(), "Array output StackRefs not supported for pure ops."
+        emitter.emit(f"_PyStackRef {outp.name}_stackref;\n")
+    stack = copy.deepcopy(stack)
+
+    storage = Storage.for_uop(stack, uop, CWriter.null(), check_liveness=False)
+    # No reference management of outputs needed.
+    for var in storage.outputs:
+        var.in_local = True
+    emitter.emit_tokens(uop, storage, None, False, is_abstract=True)
+    out.start_line()
+    # Finally, assign back the output stackrefs to symbolics.
+    for outp in uop.stack.outputs:
+        # All new stackrefs are created from new references.
+        # That's how the stackref contract works.
+        if not outp.peek:
+            out.emit(f"{outp.name} = sym_new_const_steal(ctx, PyStackRef_AsPyObjectBorrow({outp.name}_stackref));\n")
+        else:
+            out.emit(f"{outp.name} = sym_new_const(ctx, PyStackRef_AsPyObjectBorrow({outp.name}_stackref));\n")
+        storage.flush(out)
```

Comment:
I think the flush needs to go after the loop. When you have multiple outputs (not the case currently, but maybe at some point?) the storage is flushed before outputs are assigned. For example for two outputs, you end up with this:
```c
res1 = sym_new_const_steal(ctx, PyStackRef_AsPyObjectBorrow(res1_stackref));
stack_pointer[-2] = res1;
stack_pointer[-1] = res2;
res2 = sym_new_const_steal(ctx, PyStackRef_AsPyObjectBorrow(res2_stackref));
```

```suggestion
    storage.flush(out)
```

Reply:
Good catch!